### PR TITLE
Remove support for Flow comment types

### PR DESCRIPTION
**Pull request by thorn0**
## Description

See https://github.com/prettier/prettier/issues/13142#issuecomment-1239479444

Issues to check and close: https://github.com/prettier/prettier/issues?q=is%3Aopen+is%3Aissue+label%3A%22area%3Aflow+comment+types%22

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by thorn0** [comment]
Something isn't right with `babel-flow`:

**Prettier pr-13687**
[Playground link](https://deploy-preview-13687--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEZMF5MAUAJgIYwkD0AVEptZjABYCWamrmJUmEARgFZwwWKgFoKmAPIChIigEo8APkzAAvgG4AOlBAAaEBAAOMZumSgSAJysQA7gAVrCNMhAkANnZIBPVwd4rEjAAazgYAGUSAFs4ABlmKDhkADNPNDgAoNDwiKNgxIBzZBgrAFdMkDho3jgiIjq4rkKykkK4ADEIK2iyUyhilBIymAh9EEYYaI8AdRZ4NHywOAiXZlMAN3WfNzA0fxBEjKsYByDC3tT0yv40AA8Ioo84AEUyiHgrjwyDfKtjty8Ei1DyiFIeezjIxWRIwGbMIhMZAADgADL9bBkZkEjG5oXBjhtkgYAI7veBnYyuIZoURJOp1cZWOBk5jMs5tS5INLfSoZaLMErlPlPV7k5Lc64Gci8eGIxjIABM0qCzA8RQAwhBolz3B4POMyhkACrA6k8n4gDYVACSUAasAiYBhJgAgvaIjAfM8vhk1GogA)

```sh
--parser babel-flow
```

**Input:**

```jsx
const c = (data/*: /* this is an object *-/ Object */) => {};

```

**Output:**

```jsx
const c = (data: bject) => {};

```

**Comment by fisker** [comment]
AST  shows  identifier "bject"

**Comment by thorn0** [comment]
Yes, it's a bug in Babel. https://astexplorer.net/#/gist/5fdbbaf16abcc351777e4f63b57b2e77/add63d208e4e267fac75bd77ced672a97e5cfa6f

**Comment by fisker** [comment]
I thought "removing support for Flow comment types" means treat them as normal comments.  

Aren't we breaking code? (Making runnable js code not runnable)

**Comment by thorn0** [comment]
Just don't use the `flow` and `babel-flow` parsers if you want to treat them as normal comments.

**Comment by fisker** [comment]
Make sense.

**Comment by fisker** [comment]
`@babel/parser` updated #13654

**Comment by thorn0** [comment]
Treating them like normal comments will break things for people who use this syntax though. So I don't know how to proceed.

**Prettier pr-13687**
[Playground link](https://deploy-preview-13687--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEBDTBeTACgCNMB6AKiUzEwrIEoQAaECABxgEt1lRsATgIgB3AAqCEaZCGwAbEdgCe01sQHYwAazgwAytgC2cADJcocZADN5aOGo3bde9pvMBzZDAEBXeyDhDYjgAExDQk2wodx9sdzgAMQgBQ2wYbmiZbB8YCBYQAAsYQzkAdQKueDRXMDg9KUquADdKpRkwNFUQczsBGDENd1TrW38AKzQADz0POTgARR8IeBG5O1ZXAV6ZYmxguXz2AXMYUq4QmALkAA4ABg3hO1KNdhkjuF6my1YARyX4AYcaQobBoAC0FlCoXyAjgfy4sIGcWGSBsa38dkMXC8vgxswW-0sqNGrBgezOFyuSAATKSNFw5B4AMIQQwo2RyA6sHx2AAqe2BaPWICafgAklBwrA9GBjpwAIKSvQwJRzVZ2AC+GqAA)

```sh
--parser babel
```

**Input:**

```jsx
const a = (b /*: c */)
```

**Output:**

```jsx
const a = b /*: c */;

```

**Comment by thorn0** [comment]
On the other hand, if we decided to "remove support", it should mean it's okay to break that case. 🤔

**Comment by fisker** [comment]
How `hermes-eslint` handle this？

**Comment by fisker** [comment]
@gkz What do you think about this change?

**Comment by pieterv** [comment]
Removing support for this feature from prettier seems reasonable to me.

> Which might mean this syntax is considered kind of deprecated.

It roughly is, a lot of our infra already doesn't support it super well (e.g. Flows own `flow codemod` feature) and it's not so useful now that build steps are so ubiquitous even for node scripts. We will likely officially deprecate it from Flow soon.

> How hermes-eslint handle this?

`hermes-eslint` uses the hermes parser (`hermes-parser`). This is a WASM compiled version of the parser the HermesVM uses. This parser does not have any special support for Flow comments, so they stay comments (just like if you used any other parser). So they still kinda work depending on the accuracy of the prettier comment attachment logic for each comment location.

As a side note this parser works well with prettier, we use it for writing codemods: https://github.com/facebook/hermes/blob/main/tools/hermes-parser/js/hermes-transform/src/transform/transform.js#L66-L76.

**Comment by fisker** [comment]
Will the `/*:` break our whitespace/new line detect?

**Comment by thorn0** [comment]
Probably, but not more than it's already possible for the `flow` parser.

**Comment by thorn0** [comment]
Found an inconsistency between `flow` and `babel-flow`:

**Prettier pr-13687**
[Playground link](https://deploy-preview-13687--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAKAhgAgPQCpMBGmu2OuSmYx2AlCADQgQAOMAltAM7KjoBOfCAHcACvwTcU6ADZD0AT26MCfdGADWcGAGV0AWzgAZNlDjIAZjM5xlqjVu3M1JgObIYfAK42QcPQTgAE0Cgw3QoF090FzgAMQg+PXQYdgjkEHRPGAgGEAALGD1pAHU8tnhOJzA4bQlytgA3cvl0sE4lEBNrPhgRVRckiysfACtOAA9tV2k4AEVPCHgh6WtGJz5u9PNpYVzmPhMYYrZAmDzkAA4ABjXBa2LVZnT9uG6Gs0YARwX4PpZJDKcAC0piCQVyfDg3zYkL60UGSEsKx81j0bHcXhR0zmPzMiOGjBg6AIx1O5yQACZCao2NJXABhCB6BEZaTSXKeawAFWJkiRqxADW8AEkoCFYNowAdWABBMXaGDyGbLawAX1VQA)

```sh
--parser flow
```

**Input:**

```jsx
(a /* b */ /*: c */)
```

**Output:**

```jsx
(a: /* b */ c);

```

**Prettier pr-13687**
[Playground link](https://deploy-preview-13687--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAKAhgAgPQCpMBGmu2OuSmYx2AlCADQgQAOMAltAM7KjoBOfCAHcACvwTcU6ADZD0AT26MCfdGADWcGAGV0AWzgAZNlDjIAZjM5xlqjVu3M1JgObIYfAK42QcPQTgAE0Cgw3QoF090FzgAMQg+PXQYdgjkEHRPGAgGEAALGD1pAHU8tnhOJzA4bQlytgA3cvl0sE4lEBNrPhgRVRckiysfACtOAA9tV2k4AEVPCHgh6WtGJz5u9IJ0AOkAWnNpYVzmPhMYYrZAmDzkAA4ABjXBa2LVZnTTuG6Gs0YARwW8D6LEkGU4e1MQSCuT4cEBbDhfWigyQlhWPmsejY7i8mOmcyBZjRw0YMB2l2utyQACYyao2NJXABhCB6VEZaTSXKeawAFR2knRqxADW8AEkoCFYNowGdWABBKXaGDyGbLawAX01QA)

```sh
--parser babel-flow
```

**Output:**

```jsx
(a /* b */: c);

```

**Comment by fisker** [comment]
Are you going to fix it? Feel it will be difficult. How about just pretend we don't know?

**Comment by thorn0** [comment]
> Are you going to fix it? Feel it will be difficult. How about just pretend we don't know? 😄

I'm curious what's going on there, but I'll open a separate issue for it.

**Review by sosukesuzuki — APPROVED** [review]

File: changelog_unreleased/flow/13687.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#### Remove support for Flow comments (#13687 by @thorn0)
```
Comment (sosukesuzuki) [inline changelog_unreleased/flow/13687.md:1]: @thorn0 This should be marked as `[BREAKING]`?

Comment (thorn0) [reply changelog_unreleased/flow/13687.md:1]: Yes, let's do that.

Comment (thorn0) [reply changelog_unreleased/flow/13687.md:1]: Done in https://github.com/prettier/prettier/pull/13703

**Pull request merged**